### DISCUSSION TITLE
Fast hot reload in dev mode

### DISCUSSION THREAD

**Discussion by mhdaouas** [directus/directus#8190] [Feature Requests]
Hello,

Is there any way to have quick hot reload in Directus in dev mode? I tried with npx nodemon without success.

Thank you.

**Comment by joselcvarela** [comment]
Hey,
Are you talking about running locally? 
If so, and if you are talking about the app, we are using Vite for frontend and this could be what you are looking for https://vitejs.dev/guide/features.html#hot-module-replacement
If you are talking about the api, if you run this `npm run dev -- --scope directus` it should reload whenever you change a file.

If you are talking about the npm package, you can add this `    "dev": "nodemon --inspect node_modules/.bin/directus start",` to scripts in your package.json and then run `npm run dev`.

**Reply by mhdaouas** [reply]
Thanks! I will check that later!

**Reply by joselcvarela** [reply]
If this will solve your question, please mark as accepted answer 🙏

**Reply by mhdaouas** [reply]
No it didn't solve my problem. I get an error but I can't send it from where I am. Will send it in a few hours.

**Reply by mhdaouas** [reply]
After adding this line in package.json:
`"dev": "nodemon --inspect node_modules/.bin/directus start"`

and running `npm run dev`

I get:

> Couldn't register endpoint "endpoint"
> 23:21:57 ⚠️  Must use import to load ES Module: /home/myuser/Directus/my-project/extensions/endpoints/endpoint/index.js
> require() of ES modules is not supported.
> require() of /home/myuser//Directus/my-project/extensions/endpoints/endpoint/index.js from /home/myuser/Directus/my-project/node_modules/directus/dist/extensions.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
> Instead rename index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /home/myuser/Directus/my-project/package.json.

My endpoint code:
```
module.exports = function registerEndpoint(router, { services, exceptions }) {
	const { ItemsService } = services;
	const { ServiceUnavailableException } = exceptions;

	router.get('/servers', (req, res, next) => {
		const service = new ItemsService('server', { schema: req.schema, accountability: req.accountability });

		service
			.readByQuery({ sort: [{ column : 'ip', order : 'asc' }], fields: ['*'] })
			.then((results) => res.json(results))
			.catch((error) => {
				return next(new ServiceUnavailableException(error.message));
			});
	});
};
```

**Reply by joselcvarela** [reply]
Probably you are using ES Modules by either using `"type": "module"` on `package.json` or using `.mjs` file extension or even passing `--input-type=module`.
Although, Directus currently is not exported as a ES Module so my advice is to get rid of the ES Modules by revert one of the above and use the default module resolution which is CommonJS.

We are aware of this and we are pushing to go on ES Modules direction. In fact we have already started it https://github.com/directus/directus/pull/8202
On packages could be less trivial, but in API would be hard to make this changes.

**Reply by mhdaouas** [reply]
Ok on another environment I have a problem with "dirname" inside node_modules/.bin/directus. I get the error "missing ) after argument list" (2nd line of the directus file).

**Reply by mhdaouas** [reply]
Btw I'm on Windows.

**Reply by mhdaouas** [reply]
When I just run the line with "exec node ...", it says unexpected identifier for "node".

**Reply by joselcvarela** [reply]
Can you provide more information? If you take screenshots or paste the output from your terminal, could be more helpful.
I cannot find the source of the problem with the info you passed.

**Reply by mhdaouas** [reply]
Looks like I have 2 different node_modules/.bin/directus scripts on 2 different environements (I guess due to different environments: Windows and Linux) and the one which is working is the one on Linux, containing these lines:
`
#!/usr/bin/env node
require('./dist/cli/run.js')
`
It looks like the node script is not recognized on Windows even if node 14.16/or 14.17 is already installed on it.
I'm sorry I can't attach pics on Github from the Windows environment so I can't provide you with more details.

**Reply by mhdaouas** [reply]
Say I want to create my own custom layout based on the tabular layout. What should I do to develop that with hot reload and without having to `npx directus-extension build` and copy the generated index.js in the extensions/layouts folder everytime I make a change?

**Reply by joselcvarela** [reply]
I think you have only 2 options here:
1 - Using directus npm package and need to run `npx directus-extension build` everytime
2 - Run directus locally and vite.js handles hot reloads `https://docs.directus.io/contributing/running-locally/`

Am I right @nickrum ?

**Reply by mhdaouas** [reply]
It looks like the second approach is for contributing to the Directus project but what I want is just to  install the latest release with npx and add custom layouts to it (using latest stable release for personal use only).

**Reply by mhdaouas** [reply]
Maybe a solution would be to automate the build -> index.js copy process after each code change detection. What is the appropriate tool to do that? Nodemon?

**Reply by joselcvarela** [reply]
You could try with Directus locally but it seems to me the best approach is the one you mentioned:
Use nodemon to run build.

Once a npm package, the app is ready for production only, so I don't think it could be possible to support hot reload.

**Reply by nickrum** [reply]
Yes, running Directus locally in dev mode should ideally only be used when contributing.

A `--watch` option was recently added to the `directus-extension build` command. This at least automates the build process until we have a proper extension dev environment.

**Reply by joselcvarela** [reply]
Sorry for the misleading @mhdaouas 
@nickrum is our extensions master 🙂

**Comment by keesvanbemmel** [comment]
This is what works for me (the dev script):
`"dev": "yarn install; npx directus bootstrap; npx directus schema apply ./schema.yaml --yes; npx nodemon --legacy-watch --exec directus start"`

This will:

1. Install all dependencies if needed
2. Create all directus tables if needed (empty db)
3. Migrate directus to latest schema (part of bootstrapping)
4. Apply your own directus schema that is stored in root (more info)
5. Start nodemon, watching for changes in files and restarting directus if changes are detected.

Hope this helps.

**Reply by mhdaouas** [reply]
My problem is that npx directus start is very slow (on Visual Code and Windows platform). Is there a way to tweak it? Also using nodemon, I considered ignoring node_modules with the --ignore option in a hope to have at least faster restart after any code changes (at least for custom hooks) but without success.

**Reply by joselcvarela** [reply]
I think nodemon already ignores node_modules folder.
Why don't you add an entry into scripts for this: ` node node_modules/.bin/directus start`? 
Also you can try `yarn directus start`, if you have yarn.

**Reply by SeanDylanGoff** [reply]
We had the same problem, with the additional requirement that we need to run everything in docker.

Maybe that will help you aswell, it's not particularly pretty but it works.

How to use:
App extensions can just be placed uncompiled in their respective folders (so for example interfaces/test/src/index.js + interface.vue)

Api extensions have to be compiled with --watch.

If you add/remove extensions you have to restart the container - other than that no restarts should be necessarily, api extensions are reloaded in place without restarting the server and APP extensions have full HMR support.

Maybe we will document it better/make a PR sometimes, but for now its just as-is, no warranty ;)

https://github.com/cynova/directus is the repo, rc92 is on ghcr with tag -dev.
